### DISCUSSION TITLE
Convert to dictionary of pyranges objects - SCENIC+

### DISCUSSION THREAD

**Discussion by Umaarasu** [aertslab/scenicplus#100] [General]
Hi,

When I ran the example data given in Scenicplus package I didn't have any issue in this step but when I run my data, I get this error. I checked the input files and they don't seem to have any difference between my data and example data. What might be causing this error. It seems to show up when running "region_sets['DARs'][DAR] = pr.PyRanges(region_names_to_coordinates(regions))
"  Has someone faced a similar issue. Please let me know how to fix this. 

ValueError                                Traceback (most recent call last)
Cell In[232], line 15
     13 for DAR in markers_dict.keys():
     14     regions = markers_dict[DAR].index[markers_dict[DAR].index.str.startswith('chr')] #only keep regions on known chromosomes
---> 15     region_sets['DARs'][DAR] = pr.PyRanges(region_names_to_coordinates(regions))

File ~/anaconda3/envs/scenicplus/lib/python3.8/site-packages/pycistarget/utils.py:33, in region_names_to_coordinates(region_names)
     31 regiondf=pd.concat([chrom, start, end], axis=1, sort=False)
     32 regiondf.index=[i for i in region_names if ':' in i]
---> 33 regiondf.columns=['Chromosome', 'Start', 'End']
     34 return(regiondf)

File ~/anaconda3/envs/scenicplus/lib/python3.8/site-packages/pandas/core/generic.py:5915, in NDFrame.__setattr__(self, name, value)
   5913 try:
   5914     object.__getattribute__(self, name)
-> 5915     return object.__setattr__(self, name, value)
   5916 except AttributeError:
   5917     pass

File ~/anaconda3/envs/scenicplus/lib/python3.8/site-packages/pandas/_libs/properties.pyx:69, in pandas._libs.properties.AxisProperty.__set__()

File ~/anaconda3/envs/scenicplus/lib/python3.8/site-packages/pandas/core/generic.py:823, in NDFrame._set_axis(self, axis, labels)
    821 def _set_axis(self, axis: int, labels: AnyArrayLike | list) -> None:
    822     labels = ensure_index(labels)
--> 823     self._mgr.set_axis(axis, labels)
    824     self._clear_item_cache()

File ~/anaconda3/envs/scenicplus/lib/python3.8/site-packages/pandas/core/internals/managers.py:230, in BaseBlockManager.set_axis(self, axis, new_labels)
    228 def set_axis(self, axis: int, new_labels: Index) -> None:
    229     # Caller is responsible for ensuring we have an Index object.
--> 230     self._validate_set_axis(axis, new_labels)
    231     self.axes[axis] = new_labels

File ~/anaconda3/envs/scenicplus/lib/python3.8/site-packages/pandas/core/internals/base.py:70, in DataManager._validate_set_axis(self, axis, new_labels)
     67     pass
     69 elif new_len != old_len:
---> 70     raise ValueError(
     71         f"Length mismatch: Expected axis has {old_len} elements, new "
     72         f"values have {new_len} elements"
     73     )

**ValueError: Length mismatch: Expected axis has 0 elements, new values have 3 elements**

**Code:**
import pyranges as pr
from pycistarget.utils import region_names_to_coordinates
region_sets = {}
region_sets['topics_otsu'] = {}
region_sets['topics_top_3'] = {}
region_sets['DARs'] = {}
for topic in region_bin_topics_otsu.keys():
    regions = region_bin_topics_otsu[topic].index[region_bin_topics_otsu[topic].index.str.startswith('chr')] #only keep regions on known chromosomes
    region_sets['topics_otsu'][topic] = pr.PyRanges(region_names_to_coordinates(regions))
for topic in region_bin_topics_top3k.keys():
    regions = region_bin_topics_top3k[topic].index[region_bin_topics_top3k[topic].index.str.startswith('chr')] #only keep regions on known chromosomes
    region_sets['topics_top_3'][topic] = pr.PyRanges(region_names_to_coordinates(regions))
for DAR in markers_dict.keys():
    regions = markers_dict[DAR].index[markers_dict[DAR].index.str.startswith('chr')] #only keep regions on known chromosomes
    region_sets['DARs'][DAR] = pr.PyRanges(region_names_to_coordinates(regions))

**Comment by SeppeDeWinter** [comment]
Hi @Umaarasu 

One of the region sets is probably empty (i.e. no regions pass the threshold of set for calling differentially accessible regions). 

Can you check this?

Best,

Seppe

**Reply by Umaarasu** [reply]
Hi @SeppeDeWinter 

Yes, looks like there was a empty data frame in the input file. I inserted a if statement to overcome the issue. Thanks !

for DAR in markers_dict.keys():
    **if markers_dict[DAR].empty:**
          continue
    regions = markers_dict[DAR].index[markers_dict[DAR].index.str.startswith('chr')] #only keep regions on known chromosomes
    region_sets['DARs'][DAR] = pr.PyRanges(region_names_to_coordinates(regions))

**Reply by SeppeDeWinter** [reply]
Great!

Happy to help.

Best,

Seppe
